File: src/gcode/dialects.js
// dialects.js — starter dialect profiles.
//
// These are EDUCATIONAL "starter dialect profiles", not certified
// post-processors. They describe how each machine family tends to open and
// close a program and which words it expects. They make no guarantee that the
// output is safe to run on any specific machine.
//
// Each profile exposes:
//   id, label          identity
//   allowedAxes        axis letters this family understands
//   defaultUnits       'mm' | 'in'
//   supportsArcs       whether G2/G3 are idiomatic here
//   startupComment     short note for the header banner
//   headerExtras(cfg)  machine-specific header line descriptors
//   footerExtras(cfg)  machine-specific shutdown line descriptors
//
// A "line descriptor" is { cmd, words?, comment? } and is rendered by the
// generator through formatter.formatLine(). Builders are pure and DOM-free.

import { isPositiveNumber } from '../shared/numbers.js';

/** Marlin-style 3D printer: heated nozzle + bed, extrusion on the E axis. */
const marlin = {
  id: 'marlin',
  label: 'Marlin · 3D Printer',
  allowedAxes: ['X', 'Y', 'Z', 'E', 'F'],
  defaultUnits: 'mm',
  supportsArcs: false, // many Marlin builds ship without arc support
  commentStyle: ';',
  startupComment: 'heat up, home, move to safe Z',
  headerExtras(cfg) {
    const lines = [];
    const bed = isPositiveNumber(cfg.bed) ? cfg.bed : null;
    const hotend = isPositiveNumber(cfg.hotend) ? cfg.hotend : null;
    if (bed !== null) lines.push({ cmd: 'M140', words: { S: bed }, comment: 'set bed temp' });
    if (hotend !== null) lines.push({ cmd: 'M104', words: { S: hotend }, comment: 'set hotend temp' });
    if (bed !== null) lines.push({ cmd: 'M190', words: { S: bed }, comment: 'wait for bed' });
    if (hotend !== null) lines.push({ cmd: 'M109', words: { S: hotend }, comment: 'wait for hotend' });
    return lines;
  },
  footerExtras(cfg) {
    const lines = [{ cmd: 'M104', words: { S: 0 }, comment: 'hotend off' }];
    if (isPositiveNumber(cfg.bed)) lines.push({ cmd: 'M140', words: { S: 0 }, comment: 'bed off' });
    lines.push({ cmd: 'M84', comment: 'disable steppers' });
    return lines;
  },
};

/** Generic 3-axis CNC mill/router: rotating spindle, M3/M5 control. */
const genericCnc = {
  id: 'genericCnc',
  label: 'CNC Mill',
  allowedAxes: ['X', 'Y', 'Z', 'I', 'J', 'F'],
  defaultUnits: 'mm',
  supportsArcs: true,
  commentStyle: ';',
  startupComment: 'home, start spindle, move to safe Z',
  headerExtras(cfg) {
    if (!cfg.spindleOn) return [];
    return [{ cmd: 'M3', words: { S: cfg.spindleRpm ?? 0 }, comment: 'spindle on' }];
  },
  footerExtras(cfg) {
    return cfg.spindleOn ? [{ cmd: 'M5', comment: 'spindle off' }] : [];
  },
};

/** GRBL-style laser/diode engraver: beam controlled via M3/M5 + S power. */
const laserGrbl = {
  id: 'laserGrbl',
  label: 'Laser · GRBL',
  allowedAxes: ['X', 'Y', 'F'],
  defaultUnits: 'mm',
  supportsArcs: true,
  commentStyle: ';',
  startupComment: 'home, beam off, move to safe Z',
  headerExtras() {
    return [{ cmd: 'M5', comment: 'beam off' }];
  },
  footerExtras() {
    return [{ cmd: 'M5', comment: 'beam off' }];
  },
};

const REGISTRY = { marlin, genericCnc, laserGrbl };

// Backwards-compatible alias: the original app stored the CNC machine as 'cnc'.
const ALIASES = { cnc: 'genericCnc', laser: 'laserGrbl' };

/**
 * Look up a dialect profile by id (or known alias). Throws on unknown ids so
 * mistakes surface loudly rather than producing silent wrong output.
 * @param {string} id
 */
export function getDialect(id) {
  const key = REGISTRY[id] ? id : ALIASES[id];
  const dialect = REGISTRY[key];
  if (!dialect) throw new Error(`Unknown dialect: ${id}`);
  return dialect;
}

/** List all available dialect profiles. */
export function listDialects() {
  return Object.values(REGISTRY);
}

export { marlin, genericCnc, laserGrbl, REGISTRY, ALIASES };

File: python/cam_creation_studio/gcode/dialects.py
"""Starter dialect profiles.

EDUCATIONAL "starter dialect profiles", not certified post-processors. Each
describes how a machine family tends to open and close a program and which axes
it expects. No guarantee the output is safe to run on any specific machine.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any, Callable, List, Mapping

from ..shared.numbers import is_positive_number


@dataclass(frozen=True)
class LineSpec:
    """A renderable line descriptor: command word + parameter words + comment."""

    cmd: str
    words: Mapping[str, Any] = field(default_factory=dict)
    comment: str = ""


@dataclass(frozen=True)
class Dialect:
    id: str
    label: str
    allowed_axes: tuple
    default_units: str
    supports_arcs: bool
    startup_comment: str
    header_extras: Callable[[Mapping[str, Any]], List[LineSpec]]
    footer_extras: Callable[[Mapping[str, Any]], List[LineSpec]]


def _marlin_header(cfg: Mapping[str, Any]) -> List[LineSpec]:
    lines: List[LineSpec] = []
    bed = cfg.get("bed") if is_positive_number(cfg.get("bed")) else None
    hotend = cfg.get("hotend") if is_positive_number(cfg.get("hotend")) else None
    if bed is not None:
        lines.append(LineSpec("M140", {"S": bed}, "set bed temp"))
    if hotend is not None:
        lines.append(LineSpec("M104", {"S": hotend}, "set hotend temp"))
    if bed is not None:
        lines.append(LineSpec("M190", {"S": bed}, "wait for bed"))
    if hotend is not None:
        lines.append(LineSpec("M109", {"S": hotend}, "wait for hotend"))
    return lines


def _marlin_footer(cfg: Mapping[str, Any]) -> List[LineSpec]:
    lines = [LineSpec("M104", {"S": 0}, "hotend off")]
    if is_positive_number(cfg.get("bed")):
        lines.append(LineSpec("M140", {"S": 0}, "bed off"))
    lines.append(LineSpec("M84", {}, "disable steppers"))
    return lines


def _cnc_header(cfg: Mapping[str, Any]) -> List[LineSpec]:
    if not cfg.get("spindleOn"):
        return []
    return [LineSpec("M3", {"S": cfg.get("spindleRpm", 0)}, "spindle on")]


def _cnc_footer(cfg: Mapping[str, Any]) -> List[LineSpec]:
    return [LineSpec("M5", {}, "spindle off")] if cfg.get("spindleOn") else []


def _laser_header(cfg: Mapping[str, Any]) -> List[LineSpec]:
    return [LineSpec("M5", {}, "beam off")]


def _laser_footer(cfg: Mapping[str, Any]) -> List[LineSpec]:
    return [LineSpec("M5", {}, "beam off")]


marlin = Dialect(
    id="marlin",
    label="Marlin · 3D Printer",
    allowed_axes=("X", "Y", "Z", "E", "F"),
    default_units="mm",
    supports_arcs=False,
    startup_comment="heat up, home, move to safe Z",
    header_extras=_marlin_header,
    footer_extras=_marlin_footer,
)

generic_cnc = Dialect(
    id="genericCnc",
    label="CNC Mill",
    allowed_axes=("X", "Y", "Z", "I", "J", "F"),
    default_units="mm",
    supports_arcs=True,
    startup_comment="home, start spindle, move to safe Z",
    header_extras=_cnc_header,
    footer_extras=_cnc_footer,
)

laser_grbl = Dialect(
    id="laserGrbl",
    label="Laser · GRBL",
    allowed_axes=("X", "Y", "F"),
    default_units="mm",
    supports_arcs=True,
    startup_comment="home, beam off, move to safe Z",
    header_extras=_laser_header,
    footer_extras=_laser_footer,
)

_REGISTRY = {d.id: d for d in (marlin, generic_cnc, laser_grbl)}
_ALIASES = {"cnc": "genericCnc", "laser": "laserGrbl"}


def get_dialect(dialect_id: str) -> Dialect:
    """Look up a dialect by id (or known alias). Raises on unknown ids."""
    key = dialect_id if dialect_id in _REGISTRY else _ALIASES.get(dialect_id)
    dialect = _REGISTRY.get(key)
    if dialect is None:
        raise ValueError(f"Unknown dialect: {dialect_id}")
    return dialect


def list_dialects() -> List[Dialect]:
    return list(_REGISTRY.values())
